In [15]:
import pandas as pd
import numpy as np
import os
import re

In [3]:
file_list = os.listdir("./VCwallettxns/")

In [4]:
len(file_list)

468

## Total files from VC wallets: 468 wallets

### Lets explore the wallet segments

In [5]:
df = pd.DataFrame()
for file in file_list:
    filename = os.path.join("VCwallettxns",file)
    temp_df = pd.read_json(filename)
    df = pd.concat([df,temp_df])

    



In [6]:
df.reset_index(inplace=True)

In [7]:
vc_names =  pd.read_csv("VC_wallets.csv",index_col=[0])

In [8]:
vc_names

address  \
0     0xb1adceddb2941033a090dd166a462fe1c2029484   
1     0xfb86c5e6e3fdb2890d54c5f84b96d2b45c416b5d   
2     0xdf36feeb633386788fe12227b9351cbae495cba7   
3     0xdac01f49d87f128f48a8ff854a8b52ef7934c83c   
4     0xabba5bb805ccbb087872da9f49dace159a5eecda   
...                                          ...   
1281  0x500a746c9a44f68fe6aa86a92e7b3af4f322ae66   
1282  0x203520f4ec42ea39b03f62b20e20cf17db5fdfa7   
1283  0x532eedeabc0cc70f8d371ac76020e1ad9b02ceab   
1284  0x0c5a2c72c009252f0e7312f5a1ab87de02be6fbe   
1285  0xaf7eaf66b5f1fa23158784fb4c17dba90727afad   

                               label  
0                          0xb1:9484  
1         0xb1:6b5d(related_level_1)  
2         0xb1:cba7(related_level_1)  
3         0xb1:c83c(related_level_1)  
4         0xb1:ecda(related_level_1)  
...                              ...  
1281                         Voyager  
1282                         Voyager  
1283               Zee Prime Capital  
1284  Zee Prime Capital: devmons.eth  
1285    Compound Voting: Zerion:afad  

[1286 rows x 2 columns]

In [10]:
combined_df = pd.merge(df,vc_names, left_on='from',right_on='address',how="left")

In [11]:
combined_df


index  blockNumber           timeStamp  \
0           0     17423162 2023-06-06 18:24:59   
1           1     17428842 2023-06-07 13:38:35   
2           2     17472161 2023-06-13 16:08:23   
3           3     17475454 2023-06-14 03:14:35   
4           4     17539026 2023-06-23 01:30:11   
...       ...          ...                 ...   
115178     36     18390402 2023-10-20 08:39:35   
115179     37     18468169 2023-10-31 05:56:59   
115180     38     18663564 2023-11-27 14:26:11   
115181     39     18663583 2023-11-27 14:30:11   
115182      0     17878712 2023-08-09 17:03:59   

                                                     hash   nonce  \
0       0x5795a483e9580e169cb10026c5be1a5f7223930b9330...    1480   
1       0xb46bc57e2019c872e73eedfbdaee19af1ab65c8f80db...    1481   
2       0x508d1ef3ae72f496db1939f46f620e35bcee4bda5165...    1482   
3       0xc53ef65e09eaf0d18535caa836be1e8565791317a2c9...    1483   
4       0x174bd647a4a1885a00b7641fe2480252b80050882a6d...    1484   
...                                                   ...     ...   
115178  0x857b5521af93c6cb064f79c8174a95d0d8c2cbe1c85d...     542   
115179  0xf7e08b49d3ac3fad8f23f8b4d053255c020e4682f523...  552522   
115180  0x8938f93c958c3e96658f207f8008287453c638ca7319...     543   
115181  0x5b5dd0faf51ae6b9712ac19e765fd0885b1917c1db4a...     544   
115182  0xd89dee8691a640e4418de72c8efcaa34992c19375b87...     197   

                                                blockHash  transactionIndex  \
0       0x58ebd5204c569669ecd805b97babd11db75d973405c2...                99   
1       0x0550a06019bbc43118c977b91c35a739b02f3a743de4...                59   
2       0x06597ab6d3f49a19d3c9bbbba01fb2eb0cf52e70367f...                81   
3       0xc337ac533158b13d42bdba550d8805f514ea46196d11...                84   
4       0x07e570e307972f74a6cabedc11fe569bd98d8441f4d5...               320   
...                                                   ...               ...   
115178  0x0cc4bfe85e5cf9d0ecf335c18466951ac41d493d269c...               113   
115179  0x45801fcffface477759909f2b9dba2258bfb6f7a4f89...                69   
115180  0x187d3b4e0719094d97c35abf3b4404199c9ea045c3db...               141   
115181  0x0db9f4563e690e6ab1ff6521f7ed79fb287a38b7eb57...               251   
115182  0x9d507cfc503e503b995decede6bbf74b0396056e7dda...                54   

                                              from  \
0       0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
1       0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
2       0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
3       0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
4       0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
...                                            ...   
115178  0x9ec9c16e67e089d7050519d95975b7b807a5b289   
115179  0xb8001c3ec9aa1985f6c747e25c28324e4a361ec1   
115180  0x9ec9c16e67e089d7050519d95975b7b807a5b289   
115181  0x9ec9c16e67e089d7050519d95975b7b807a5b289   
115182  0x62ffb5b8f6c6e3dc949f4e7b163831fe6d2ed423   

                                                to         value  ...  \
0       0x9d90669665607f08005cae4a7098143f554c59ef  7.770000e+14  ...   
1       0x00000000000000adc04c56bf30ac9d3c0aaf14dc  7.350000e+17  ...   
2       0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad  1.000000e+17  ...   
3       0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad  5.000000e+17  ...   
4       0xe4cfae3aa41115cb94cff39bb5dbae8bd0ea9d41  0.000000e+00  ...   
...                                            ...           ...  ...   
115178  0x40a50cf069e992aa4536211b23f286ef88752187  3.000000e+19  ...   
115179  0x9ec9c16e67e089d7050519d95975b7b807a5b289  3.921719e+17  ...   
115180  0x40a50cf069e992aa4536211b23f286ef88752187  1.500000e+19  ...   
115181  0x2260fac5e5542a773aa44fbcfedf7c193bc2c599  0.000000e+00  ...   
115182  0x41318419cfa25396b47a94896ffa2c77c6434040  3.893901e+16  ...   

        txreceipt_status                                              input  \
0                      1

Number of unique addreses which have function selectors in call data


In [16]:
combined_df[['VCname', 'VCname_level']] = combined_df['label'].str.extract(r'([^:]+):(.+)')

In [23]:
combined_df['functionName'].value_counts(normalize=True)

functionName
                                                                                                                                              0.460085
transfer(address _to, uint256 _value)                                                                                                         0.270543
sendMultiSig(address toAddress, uint256 value, bytes data, uint256 expireTime, uint256 sequenceId, bytes signature)                           0.094919
sendMultiSigToken(address toAddress, uint256 value, address tokenContractAddress, uint256 expireTime, uint256 sequenceId, bytes signature)    0.078536
approve(address _spender, uint256 _value)                                                                                                     0.015037
                                                                                                                                                ...   
unstakeClaimTokens_new(uint256 unbondNonce)                                      

In [38]:
df_w_func = combined_df[combined_df['functionName']!=""]
df_wo_func = combined_df[combined_df['functionName']==""]


VCname
0xb1                        8
7 OClock Capital            7
Amber                     318
Arca                        4
Arrington XRP Capital       2
                         ... 
Three Arrows Capital     1009
Tiantian Kullander          3
Varys Capital               4
Zee Prime Capital          31
hindsightcapital            4
Length: 70, dtype: int64

In [1]:
combined_df.groupby(['contractAddress','gasUsed']).count()

NameError: name 'combined_df' is not defined

In [39]:
df_wo_func

index  blockNumber           timeStamp  \
15         15     18066163 2023-09-04 22:40:35   
16         16     18066174 2023-09-04 22:42:47   
17         17     18081602 2023-09-07 02:34:47   
18         18     18081617 2023-09-07 02:37:47   
19         19     18081628 2023-09-07 02:39:59   
...       ...          ...                 ...   
115138      5     18578956 2023-11-15 18:04:47   
115140      0     17463515 2023-06-12 10:54:23   
115141      1     17463518 2023-06-12 10:54:59   
115179     37     18468169 2023-10-31 05:56:59   
115182      0     17878712 2023-08-09 17:03:59   

                                                     hash   nonce  \
15      0xc9812716afb10cc1160e4d60b3cc70c40ce804bdca10...       1   
16      0x031a75e3a0d9fd9f32e45eadd503819fbd653cb624eb...    1495   
17      0xf2ba96d273011a4681c5e3908ae8ffb1f7df73ebba56...     112   
18      0x12daa476ea3134c7eab547375011c613884c8d22fbe2...    1496   
19      0x4faab6ac9c08248152d24ec84ba308c499677cc6a23f...    1497   
...                                                   ...     ...   
115138  0x5e3eb00833db274843ed8cd9bb025b092f13c986794a...    1325   
115140  0x2c3a13b66ca88ad33601ef77b7b0f3f13f5a9db59b66...    1434   
115141  0x68fe8ca3fcc86fe14e70074d9dfd738b34d4182e53f2...    1435   
115179  0xf7e08b49d3ac3fad8f23f8b4d053255c020e4682f523...  552522   
115182  0xd89dee8691a640e4418de72c8efcaa34992c19375b87...     197   

                                                blockHash  transactionIndex  \
15      0xfa3d0994efed4ade2377de86a833f320ee2500aab3b2...                11   
16      0xd17cba140ca66d6299b1705569663074b8036addee9f...                60   
17      0xc81b2bb62a9f425ea841d46646cd9b6401ec64d0fc22...               117   
18      0x18447938f9a3d1a399ff964053f4740bc4c5a70c9095...                89   
19      0x7a0097b13f6f1a98c19704be984be6b9b01cd7149f55...                55   
...                                                   ...               ...   
115138  0x4eb10ef73d8356277385f1a29d521483cc32aaa4c5db...               139   
115140  0xe9e939814942cd3d6e7ca31bab8f19fddc39843284d9...                98   
115141  0xa03ec9d681212af2386f6d3ddd0ad9b8e3c9f1a2a2b0...               108   
115179  0x45801fcffface477759909f2b9dba2258bfb6f7a4f89...                69   
115182  0x9d507cfc503e503b995decede6bbf74b0396056e7dda...                54   

                                              from  \
15      0xc2583a3bc2df23b3f678e1fbeaa3cfd95891d8d6   
16      0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
17      0x82a8d7d1371fe53dca49214116f0fef846bab08b   
18      0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
19      0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f   
...                                            ...   
115138  0xe8d49dc32df7d446d3f17b8205f526754f374bc8   
115140  0x8dec37c0034bc3abbf4e204b27bef628eb7cc944   
115141  0x8dec37c0034bc3abbf4e204b27bef628eb7cc944   
115179  0xb8001c3ec9aa1985f6c747e25c28324e4a361ec1   
115182  0x62ffb5b8f6c6e3dc949f4e7b163831fe6d2ed423   

                                                to         value  ...  \
15      0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f  1.250000e+20  ...   
16      0x78ef85b229ef3143e18cec168554b6395d877603  1.250000e+20  ...   
17      0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f  1.252500e+20  ...   
18      0xc2583a3bc2df23b3f678e1fbeaa3cfd95891d8d6  1.250000e+20  ...   
19      0xc2583a3bc2df23b3f678e1fbeaa3cfd95891d8d6  5.137946e+14  ...   
...                                            ...           ...  ...   
115138  0xefe5a4a033b19497114135cb08bb67c01f2dc39d  7.033740e+17  ...   
115140  0x8dec37c0034bc3abbf4e204b27bef628eb7cc944  0.000000e+00  ...   
115141  0x41574a52b3dc47c8c68966fe3595266cbd8896b2  5.980274e+18  ...   
115179  0x9ec9c16e67e089d7050519d95975b7b807a5b289  3.921719e+17  ...   
115182  0x41318419cfa25396b47a94896ffa2c77c6434040  3.893901e+16  ...   

        contractAddress  cumulativeGasUsed  gasUsed  confirmations methodId  \
15                      

In [21]:
df_dict = df.to_dict(orient="records")

In [22]:
df_dict[0]

{'index': 0,
 'blockNumber': 17423162,
 'timeStamp': Timestamp('2023-06-06 18:24:59'),
 'hash': '0x5795a483e9580e169cb10026c5be1a5f7223930b9330ddb0ad09e884ae4dc524',
 'nonce': 1480,
 'blockHash': '0x58ebd5204c569669ecd805b97babd11db75d973405c201bcd4ad000d9bfb68cb',
 'transactionIndex': 99,
 'from': '0x3e95fef1176acf5e5d2ef67d9c856e4ecac73e1f',
 'to': '0x9d90669665607f08005cae4a7098143f554c59ef',
 'value': 777000000000000.0,
 'gas': 116900,
 'gasPrice': 25726629114,
 'isError': 0,
 'txreceipt_status': 1,
 'input': '0xefef39a10000000000000000000000000000000000000000000000000000000000000001',
 'contractAddress': '',
 'cumulativeGasUsed': 7408058,
 'gasUsed': 110595,
 'confirmations': 1277729,
 'methodId': '0xefef39a1',
 'functionName': 'purchase(uint256 _price)'}